### PR TITLE
Add proposal for Provider Hook

### PR DESCRIPTION
**Pull request by s-sen**
Signed-off-by: ssen <mailssen@gmail.com>

### CODE REVIEW

File: 005-OFEP-provider-hook.md
Comment (beeme1mr) [inline 005-OFEP-provider-hook.md:43]: Could you elaborate on why the provider hooks would run before the invocation hooks? I would have expected the provider hooks to run last in the before hook and first in subsequent hooks. That would allow provider developers to define hooks that are guaranteed to run right before flag evaluation.

**Review by beeme1mr — COMMENTED** [review]
Thanks for creating this proposal. I think provider hooks would add additional flexibility for provider developers without introducing new concepts.

My only concern is hook ordering. I've elaborated more on this in the proposal itself.

File: 005-OFEP-provider-hook.md
Comment (s-sen) [reply 005-OFEP-provider-hook.md:43]: In the before stage, the Provider hook can run after the client - I will modify the current order for this.

But for the after stage, one use case is that provider can do some operation that can be overridden by the invocation hook. If the provider hook runs before the invocation hook, the one way the application developer can turn off the default behavior is to use hook hints so that even if the provider hook runs after, it will turn off the default behavior using hook hints. 

What is the use case of the Provider hook to run first in the after stage? If the use case is strong we can mandate that the provider hook can run before the invocation hook - in that case for specific overriding use cases the provider can supply hook hints.

File: 005-OFEP-provider-hook.md
Comment (justinabrahms) [inline 005-OFEP-provider-hook.md:43]: I dislike the lack of symmetry here. Would it be reasonable to have providers run their before hooks first? I'd love to maintain a simple "the after is the reverse order of before" if we can.

Comment (s-sen) [reply 005-OFEP-provider-hook.md:43]: In the 'before' stage, IMO, the Provider should run last. This is because in this stage, the context is mutable and the Provider should be responsible for examining the final form of the context and then run the evaluation.
I am not clear on why the hooks should run the reverse order in the 'after' stage. I am curious about the use cases that rationalizes this.

**Comment by toddbaert** [comment]
Just to clarify: **is there any reason that we couldn't use our existing hooks abstraction for this?** and just attach them at the provider level as specified. My main concern is adding new abstractions... if provider hooks use the same hook abstraction as our existing hooks, I'm more comfortable with this proposal.

I also feel that provider hooks should be "innermost" hooks - they should run last in before, after the application author has run their hooks, and first in error, after, and finally.

File: 005-OFEP-provider-hook.md
Comment (toddbaert) [reply 005-OFEP-provider-hook.md:43]: > In the 'before' stage, IMO, the Provider should run last. This is because in this stage, the context is mutable and the Provider should be responsible for examining the final form of the context and then run the evaluation.

Agreed.

> I am not clear on why the hooks should run the reverse order in the 'after' stage. I am curious about the use cases that rationalizes this.

Especially for global hooks, having them run in stack order makes a lot of sense - for example, with the OpenTelementry hook, we want to start the span as early as possible and close it at the very end, after all hooks have run, for maximum visibility.

I think in all cases, because these are provider hooks, they should run as close to the provider as possible, so:

```
before: API, Client, Invocation, Provider
after/error/finally: Provider, Invocation, Client, API
```

Comment (skyerus) [reply 005-OFEP-provider-hook.md:43]: Food for thought, if the ordering of hooks is
```
before: API, Client, Invocation, Provider
after/error/finally: Provider, Invocation, Client, API
```
the only hook that would allow the provider customisability beyond what is already available is the `finally` hook.
A provider already has direct control over what happens immediately before and after a flag call. Whereas there is currently no way for a provider to hook (`finally`) into the execution post success/failure.

Comment (toddbaert) [reply 005-OFEP-provider-hook.md:43]: >A provider already has direct control over what happens immediately before and after a flag call. Whereas there is currently no way for a provider to hook (finally) into the execution post success/failure.

I think this is good, TBH, that's the business of the application author.

What there currently is _not_, is the ability to easily run arbitrary code before/after flag evaluation in a provider, which can lead to a lot of duplicated code in a provider for each flag resolution type (the context-transformer was a way to avoid this, but it has not been widely adopted and only handled one specific use case).

Comment (s-sen) [reply 005-OFEP-provider-hook.md:43]: I have modified the after hook order as discussed above. 
I like the idea of running the Provider hook  after the invocation in the 'finally' phase to but this will require more logic in the SDK to ensure the following order for finally: Invocation, Provider, Client, API.

**Comment by toddbaert** [comment]
@s-sen:

>Just to clarify: is there any reason that we couldn't use our existing hooks abstraction for this? and just attach them at the provider level as specified? My main concern is adding new abstractions... if provider hooks use the same hook abstraction as our existing hooks, I'm more comfortable with this proposal.

Can you add any clarification on this? To be clear, I'm just making sure that a provider hook would use the same hook interface that already exists... it seems like that's the case.

**Comment by toddbaert** [comment]
I'd also like to point out I think this makes the `contextTransformer` concept redundant. Since before hooks can transform context, a provider before hook is a great place to do such transformation. I think that if we add provider hooks, we should completely remove the `contextTransformer` concept.

cc @thomaspoignant @beeme1mr @weyert

**Comment by s-sen** [comment]
> Can you add any clarification on this? To be clear, I'm just making sure that a provider hook would use the same hook interface that already exists... it seems like that's the case

Yes, there is no change proposed in the current hook interface. The Provider Hook implementation must use the same interface that is currently defined.

But the application developer should not have to register/add the hook. During running the hooks, the SDK should query the Provider implementation for any provider hooks and if it is provided run the Provider hook in the order that the standard mandates.

**Review by toddbaert — APPROVED** [review]
After above discussion, I would support a change to the spec outlining this.

Interested in getting more feedback from others though.

**Review by beeme1mr — APPROVED** [review]
Looks good to me.

**Comment by thomaspoignant** [comment]
> 'd also like to point out I think this makes the `contextTransformer` concept redundant. Since before hooks can transform context, a provider before hook is a great place to do such transformation. I think that if we add provider hooks, we should completely remove the `contextTransformer` concept.

100% agree, keeping both will be confusing.

**Review by thomaspoignant — APPROVED** [review]

**Review by justinabrahms — APPROVED** [review]

**Comment by justinabrahms** [comment]
I really like that this gets rid of context transformers. Thanks, @s-sen!!

**Comment by toddbaert** [comment]
@s-sen I'm going to merge this if you agree.

**Review by InTheCloudDan — APPROVED** [review]

**Comment by beeme1mr** [comment]
@s-sen, it looks like we have consensus. Please update the OFEP status to `APPROVED` and I'll merge it in.

Once the OFEP has been merged, feel free to open a PR in the spec repo with the required changes and reference this PR.

Thanks for driving this spec enhancement.

**Comment by s-sen** [comment]
Sorry - I could not do this earlier. I do not need to do any additional changes in the document I assume? 

The next step is to open a PR in the spec repo?

**Comment by beeme1mr** [comment]
@s-sen, nothing else is required here. The next step would be to open a PR in the spec repo and reference this PR.

Thanks for driving this topic. I think it's a great enhancement.

**Pull request merged**
